File: __prologue_server/ecs/manager.go
/* */

package ecs

import (
	"fmt"
	"prologue/ecs/comps"
)

type ComponentManager[T any] struct {
	Components map[int]T
}

func (cm *ComponentManager[T]) AddComponent(id int, component T) {
	cm.Components[id] = component
}

func (cm *ComponentManager[T]) GetComponent(id int) (T, bool) {
	var zero T
	component, exists := cm.Components[id]
	if !exists {
		return zero, false
	}
	return component, true
}

func (cm *ComponentManager[T]) RemoveComponent(id int) {
	delete(cm.Components, id)
}

type ECS struct {
	counterID int
	Positions ComponentManager[comps.Position]
	Infos     ComponentManager[comps.Info]
	Healths   ComponentManager[comps.Health]
	tags      map[int]map[string]struct{}
}

func NewECS() *ECS {
	return &ECS{
		counterID: 1,
		Positions: ComponentManager[comps.Position]{
			Components: make(map[int]comps.Position),
		},
		Infos: ComponentManager[comps.Info]{
			Components: make(map[int]comps.Info),
		},
		Healths: ComponentManager[comps.Health]{
			Components: make(map[int]comps.Health),
		},
		tags: make(map[int]map[string]struct{}),
	}
}

func (e *ECS) NewEntity() int {
	id := e.counterID
	e.counterID++
	e.tags[id] = make(map[string]struct{})
	return id
}

func (e *ECS) AddTag(id int, tag string) {
	_, exists := e.tags[id]
	if !exists {
		return
	}
	e.tags[id][tag] = struct{}{}
}

func (e *ECS) RemoveTag(id int, tag string) {
	_, exists := e.tags[id]
	if !exists {
		return
	}
	delete(e.tags[id], tag)
}

func (e *ECS) HasTag(id int, tag string) bool {
	_, exists := e.tags[id][tag]
	return exists
}

func (e *ECS) GetEntitiesWithTag(tag string) []int {
	var result []int
	for id, tags := range e.tags {
		_, exists := tags[tag]
		if exists {
			result = append(result, id)
		}
	}
	return result
}

func (e *ECS) PrintEntity(id int) {
	fmt.Printf("Entity %d:\n", id)

	pos, exists := e.Positions.GetComponent(id)
	if exists {
		fmt.Printf("  Position: %+v\n", pos)
	}

	info, exists := e.Infos.GetComponent(id)
	if exists {
		fmt.Printf("  Info: %+v\n", info)
	}

	health, exists := e.Healths.GetComponent(id)
	if exists {
		fmt.Printf("  Health: %+v\n", health)
	}

	tags := []string{}
	for tag := range e.tags[id] {
		tags = append(tags, tag)
	}
	fmt.Println("  Tags:", tags)
}

File: __prologue_client/core/ecs/manager.js
/**/

class ECS {
  constructor() {
    this.counterID = 1;
    this.entities = new Map();   // id -> Set(componentTypes)
    this.components = new Map(); // compType -> Map(id, data)
    this.tags = new Map();       // id -> Set(tags)
  }

  newEntity() {
    const id = this.counterID++;
    this.entities.set(id, new Set());
    this.tags.set(id, new Set());
    return id;
  }

  list() {
    for (const [id, components] of this.entities.entries()) {
      console.log(id, [...components], [...(this.tags.get(id) || [])]);
    }
  }

  addTag(id, tag) {
    if (!this.tags.has(id)) {
      return;
    }
    this.tags.get(id).add(tag);
  }

  removeTag(id, tag) {
    if (!this.tags.has(id)) {
      return;
    }
    this.tags.get(id).delete(tag);
  }

  hasTag(id, tag) {
    return this.tags.has(id) && this.tags.get(id).has(tag);
  }

  getEntitiesWithTag(tag) {
    const result = [];
    for (const [id, tags] of this.tags) {
      if (tags.has(tag)) result.push(id);
    }
    return result;
  }

  getEntitiesWithComp(compType) {
    const result = [];
    for (const [id, components] of this.entities) {
      if (components.has(compType)) result.push(id);
    }
    return result;
  }

  addComponent(id, compType, comp) {
    if (!this.entities.has(id)) {
      return;
    }
    this.entities.get(id).add(compType);
    if (!this.components.has(compType)) {
      this.components.set(compType, new Map());
    }
    this.components.get(compType).set(id, comp);
  }

  removeComponent(id, compType) {
    if (!this.entities.has(id) || !this.components.has(compType)) {
      return;
    }
    this.entities.get(id).delete(compType);
    this.components.get(compType).delete(id);
  }

  printEntity(id) {
    if (!this.entities.has(id)) {
      console.log(`Entity ${id} does not exist.`);
      return;
    }

    console.log(`Entity ${id}:`);
    console.log("  Components:", [...this.entities.get(id)]);
    console.log("  Tags:", [...(this.tags.get(id) || [])]);

    for (const comp of this.entities.get(id)) {
      const data = this.components.get(comp)?.get(id);
      console.log(`  ${comp}:`, data);
    }
  }
}

export {
  ECS
};
